### PR TITLE
Added payment plugin

### PR DESCRIPTION
**Pull request by shreypandey**

### CODE REVIEW

**Review by sameersegal — APPROVED** [review]

File: plugins/payment-plugin/payment_plugin/payment.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+    response = requests.request("POST", url, headers=headers, data=payload)
+    response = json.loads(response.text)
+
+    payment_url = response["short_url"]
+    return payment_url
+
+
+class PaymentPluginReturnStatus(Enum):
+    SUCCESS = "success"
```
Comment (sameersegal) [inline plugins/payment-plugin/payment_plugin/payment.py]: I think the ENUM values need to be capital letters to be compatible with the YAML

File: plugins/payment-plugin/payment_plugin/payment.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+    def on_enter_send_payment_success_confirmation(self):
+        self.status = Status.WAIT_FOR_ME
+        self.return_status = self.return_status_values.SUCCESS
+        self.send_message(
+            FSMOutput(
+                intent=FSMIntent.SEND_MESSAGE,
+                message=Message(
+                    message_type=MessageType.TEXT,
+                    text=TextMessage(body="Payment Received!"),
```
Comment (sameersegal) [inline plugins/payment-plugin/payment_plugin/payment.py:155]: If my plugin returns multiple output variables, how will that work?

File: plugins/payment-plugin/payment_plugin.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+name: PaymentPlugin
+description: |
+    The name of the plugin is payment. This plugin takes payment amount to make a payment for a specific reason.
+
+CREDENTIALS:
```
Comment (sameersegal) [inline plugins/payment-plugin/payment_plugin.yaml:5]: Why did you change the YAML?

File: plugins/payment-plugin/payment_plugin/payment.py (hunk 1 above)
Comment (shreypandey) [reply plugins/payment-plugin/payment_plugin/payment.py]: @sameersegal codegen should access enum by name, not by value. In that case, value does not matter. 
For convention, we can have same string for names and values.

File: plugins/payment-plugin/payment_plugin/payment.py (hunk 2 above)
Comment (shreypandey) [reply plugins/payment-plugin/payment_plugin/payment.py:155]: Plugin FSM returns a dict, so it can incorporate multiple return values. Check the YAML.

File: plugins/payment-plugin/payment_plugin.yaml (hunk 3 above)
Comment (shreypandey) [reply plugins/payment-plugin/payment_plugin.yaml:5]: To make it understandable by codegen.

Comment (sameersegal) [reply plugins/payment-plugin/payment_plugin.yaml:5]: NL2DSL doesn't understand credentials. Let's stick to inputs, outputs and error codes only

File: plugins/payment-plugin/payment_plugin/payment.py (hunk 2 above)
Comment (sameersegal) [reply plugins/payment-plugin/payment_plugin/payment.py:155]: Where is the output variable getting set?

File: plugins/payment-plugin/payment_plugin.yaml (hunk 3 above)
Comment (shreypandey) [reply plugins/payment-plugin/payment_plugin.yaml:5]: It doesn't matter. Credentials will be passed by codegen during plugin initialization. 
Here is an example of plugin initialization with credentials, in the calling FSM:

```python
self.plugins: Dict[str, AbstractFSM] = {
     "payment": PaymentPlugin(
            send_message=send_message,
            credentials={
                    "RAZORPAY_CLIENT_ID": self.credentials.get("RAZORPAY_CLIENT_ID"),
                    "RAZORPAY_CLIENT_SECRET": self.credentials.get("RAZORPAY_CLIENT_SECRET"),
            },
     ),
}
```

Codegen can handle this initialisation.

File: plugins/payment-plugin/payment_plugin/payment.py (hunk 2 above)
Comment (shreypandey) [reply plugins/payment-plugin/payment_plugin/payment.py:155]: In line 93, we are defining the variable names whose values will be converted to output when plugin FSM ends. This part is handled by `AbstractFSM`.

File: plugins/payment-plugin/payment_plugin.yaml (hunk 3 above)
Comment (sameersegal) [reply plugins/payment-plugin/payment_plugin.yaml:5]: But not NL2DSL. We will solve this later. I don't want to disturb NL2DSL. At the moment take all arguments through the function call and not through constructor.

Comment (shreypandey) [reply plugins/payment-plugin/payment_plugin.yaml:5]: NL2DSL doesn't have to change to incorporate credentials. Since credentials names are already defined in this yaml description, codegen can generate the above code snippet. We don't need to change NL2DSL.

**Review by sameersegal — APPROVED** [review]
